_source/_posts/2023-07-27-spring-graphql-react.adoc

```diff
@@ -0,0 +1,1857 @@
+      });
+    }
+  }, [isAuthenticated, isLoading, loginWithRedirect]);
+
+  if (isLoading) {
+    return <Loader />;
+  }
+  if (error) {
+    return <div>Oops... {error.message}</div>;
+  }
+  return <>{isAuthenticated && props.children}</>;
+};
+
+export default AuthenticationGuard;
+----
+
+The `AuthenticationGuard` component will be used to protect pages that require authentication, redirecting to the Auth0 universal login. Protect the index page by wrapping its content in the `AuthenticationGuard` component:
+
+.app/page.tsx
+[source, tsx]
+----
+{% raw %}
+import AuthenticationGuard from "@/components/authentication/AuthenticationGuard";
+import HomePage from "./HomePage";
+
+const Page = () => {
+  return (
+    <AuthenticationGuard>
+      <HomePage></HomePage>
+    </AuthenticationGuard>
+  );
+};
+
+export default Page;
+{% endraw %}
+----
+
+=== Call the API server with an accessToken
+
+Add th file `src/services/auth.tsx` with the following code:
+
+.auth.tsx
+[source, tsx]
+----
+import backendAPI from "./base";
+
+let requestInterceptor: number;
+let responseInterceptor: number;
+
+export const clearInterceptors = () => {
+  backendAPI.interceptors.request.eject(requestInterceptor);
+  backendAPI.interceptors.response.eject(responseInterceptor);
+};
+
+export const setInterceptors = (accessToken: String) => {
+
+  clearInterceptors();
+
+  requestInterceptor = backendAPI.interceptors.request.use(
+    // @ts-expect-error
+    function (config) {
+      return {
+        ...config,
+        headers: {
+          ...config.headers,
+          Authorization: `Bearer ${accessToken}`,
+        },
+      };
+    },
+    function (error) {
+      console.log("request interceptor error", error);
+      return Promise.reject(error);
+    }
+  );
+};
+----
+
+Add the file `src/hooks/useAccessToken.tsx` with the following content:
+
+.useAccessToken.tsx
+[source, tsx]
+----
+import { setInterceptors } from "@/services/auth";
+import { useAuth0 } from "@auth0/auth0-react";
+import { useCallback, useState } from "react";
+
+export const useAccessToken = () => {
+  const { isAuthenticated, getAccessTokenSilently } = useAuth0();
+  const [accessToken, setAccessToken] = useState("");
+
+  const saveAccessToken = useCallback(async () => {
+    if (isAuthenticated) {
+      try {
+        const tokenValue = await getAccessTokenSilently();
+        if (accessToken !== tokenValue) {
+          setInterceptors(tokenValue);
+          setAccessToken(tokenValue);
+        }
+      } catch (err) {
+        // Inactivity timeout
+        console.log("getAccessTokenSilently error", err);
+      }
+    }
+  }, [getAccessTokenSilently, isAuthenticated, accessToken]);
+
+  return {
+    saveAccessToken,
+  };
+};
+----
+
+The hook will call Auth0 `getAccessTokenSilently` and will trigger a token refresh if the accessToken is expired. Then it will update axios interceptors to set the updated bearer token value in the request headers.
```

Comment:
```suggestion
The hook will call Auth0 `getAccessTokenSilently()` and will trigger a token refresh if the accessToken is expired. Then it will update axios interceptors to set the updated bearer token value in the request headers.
```